File: python/st_discord_bot.py
import logging

import nextcord
from nextcord.ext import commands
from nextcord.utils import get
from typing import Optional

bot: commands.Bot = commands.Bot(
    command_prefix="_",
    help_command=None,
    member_cache_flags=nextcord.MemberCacheFlags.none(),
)


@bot.event
async def on_ready() -> None:
    logging.info(f"{bot.user} is ready!")
    logging.info(bot.guilds)


@bot.event
async def on_voice_state_update(
    member: nextcord.Member,
    before: nextcord.VoiceState,
    after: nextcord.VoiceState,
) -> None:
    if member.bot:
        return
    if after.channel and after.channel.id in [1234, 5678]:
        role: Optional[nextcord.Role] = get(member.guild.roles, name="studying")
        if role:
            await member.add_roles(role, atomic=True)
    if before.channel and before.channel.id in [1234, 5678]:
        role = get(member.guild.roles, name="studying")
        if role:
            await member.remove_roles(role, atomic=True)


bot.run("someKey")

File: typescript/st_discord_bot.ts
import {
  Client,
  GatewayIntentBits,
  GuildMember,
  VoiceState,
  Role,
  Guild,
} from "discord.js";

const bot = new Client({
  intents: [
    GatewayIntentBits.Guilds,
    GatewayIntentBits.GuildMembers,
    GatewayIntentBits.GuildVoiceStates,
  ],
});

bot.once("ready", () => {
  console.log(`${bot.user?.tag} is ready!`);
  console.log(bot.guilds.cache);
});

bot.on("voiceStateUpdate", async (before: VoiceState, after: VoiceState) => {
  const guild: Guild | null = after.guild ?? before.guild;
  if (!guild) return;

  const member: GuildMember | undefined =
    guild.members.cache.get(after.id) || guild.members.cache.get(before.id);

  // If member is undefined or a bot we exit the function.
  if (!member || member.user.bot) return;

  const roleName = "studying";
  const channelIds = [1234, 5678];

  if (after.channel && channelIds.includes(Number(after.channel.id))) {
    const role: Role | undefined = guild.roles.cache.find(
      (r) => r.name === roleName
    );
    if (role) {
      await member.roles.add(role, "User joined studying voice channel");
    }
  }

  if (before.channel && channelIds.includes(Number(before.channel.id))) {
    const role: Role | undefined = guild.roles.cache.find(
      (r) => r.name === roleName
    );
    if (role) {
      await member.roles.remove(role, "User left studying voice channel");
    }
  }
});

bot.login("someKey");
